### PR TITLE
Create Link between Notes to Nodes and Transactions 

### PR DESCRIPTION
**Pull request by Delphinus8821**
### Description of the Change
User Notes now store which nodes and transactions were selected when the note was created if the user chooses to apply the note to the current graph selection. 

Selected notes fom graph selection

Also added the ability for users to change the selection for a specific note from the edit dialog. 3 radio buttons have been added to the bottom of the edit dialog window where the user is able to choose to keep the original selection, add what is currently selected on the graph to the note selection, or remove what is currently selected on the graph from the note selection. 
If the note was applied to the entire graph and the user then edits the note and adds nodes and/or transactions to the selection then the note will no longer be applied to the entire graph and only to those nodes/transactions. Or, if the user were to remove all of the nodes and transactions the note was specifically applied to, the note would then be applied to the entire graph. 

EditNote

A right click menu added to user notes for users to select the related nodes and transactions on the graph for that note. This menu was only added to user notes for now as the only option is select on graph

A selection filter was also created so that when in the selection filter, notes relating to the graph elements currently selected appear. E.g. if nothing is selected, only the notes that are applied to the entire graph appear. If nodes and/or transactions are selected which are related to a user note that has been created that note appears as well as all of the graph notes. This only applies to user notes and the other filters are not able to be selected while the selected filter is. 

### Alternate Designs
None realised.

### Why Should This Be In Core?
Allows for users to assign nodes and transactions to a specific note. This lets users write notes about specific graph elements. 

### Possible Drawbacks
None realised. 

### Applicable Issues
#1004

### CODE REVIEW

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 1]
```diff
@@ -89,20 +102,29 @@
...
 
     private static final String NOTES_VIEW_ICON = "resources/notes-view.png";
 
     private final Object LOCK = new Object();
 
     private final String fontStyle = String.format("-fx-font-size:%d;", FontUtilities.getApplicationFontSize());
 
+    private final List<Integer> nodesSelected = new ArrayList();
+    private final List<Integer> transactionsSelected = new ArrayList();
+    private Boolean applySelected;
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:124]: change this to the primitive type `boolean`

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 2]
```diff
@@ -157,6 +179,18 @@ public void onChanged(final ListChangeListener.Change event) {
...
+        // Checkbox to apply note to selection
+        final CheckBox applyToSelection = new CheckBox("Apply note to graph selection");
+        applyToSelection.setSelected(true);
+        applySelected = true;
+        applyToSelection.selectedProperty().addListener((ov, oldVal, newVal) -> {
+            if (applyToSelection.isSelected() == true) {
+                applySelected = true;
+            } else {
+                applySelected = false;
+            }
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:191]: can reduce this if-else to `applySelected = applyToSelection.isSelected()`

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 3]
```diff
@@ -183,14 +217,60 @@ public void onChanged(final ListChangeListener.Change event) {
...
-                                Long.toString(ZonedDateTime.now().toInstant().toEpochMilli()),
-                                titleField.getText(),
-                                contentField.getText(),
-                                true
-                        ));
+                        if (applySelected) {
+                            // Get selected nodes from the graph
+                            List<Integer> selectedNodes = new ArrayList();
+                            // Get selected transactions from the graph
+                            List<Integer> selectedTransactions = new ArrayList();
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:224]: both final + add diamond operator

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 4]
```diff
@@ -333,15 +419,34 @@ protected void setFilters(final List<String> selectedFilters) {
...
                     if ((selectedFilters.contains(USER_NOTES_FILTER) && entry.isUserCreated())
                             || (selectedFilters.contains(AUTO_NOTES_FILTER) && !entry.isUserCreated())) {
                         notesToRender.add(entry);
+                    } else if (selectedFilters.contains(SELECTED_FILTER) && entry.isUserCreated()) {
+                        // if no nodes or transactions are selected, show notes applied to the whole graph 
+                        if (entry.isGraphAttribute()) {
+                            notesToRender.add(entry);
+                        }
+                        // Show notes related to the selected nodes
+                        for (int node : nodesSelected) {
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:438]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 5]
```diff
@@ -333,15 +419,34 @@ protected void setFilters(final List<String> selectedFilters) {
...
+                            if (entry.getNodesSelected() != null && entry.getNodesSelected().contains(node) && !notesToRender.contains(entry)) {
+                                notesToRender.add(entry);
+                            }
+                        }
+                        // Shows notes related to the selected transactions 
+                        for (int transaction : transactionsSelected) {
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:444]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 6]
```diff
@@ -556,4 +730,107 @@ protected void closeEdit() {
...
+     */
+    public void addToSelectedElements(final NotesViewEntry noteToEdit) {
+        updateSelectedElements();
+
+        if (!nodesSelected.isEmpty()) {
+            if (noteToEdit.isGraphAttribute()) {
+                noteToEdit.setGraphAttribute(false);
+            }
+            final List<Integer> originalNodes = noteToEdit.getNodesSelected();
+            for (int node : nodesSelected) {
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:781]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 7]
```diff
@@ -556,4 +730,107 @@ protected void closeEdit() {
...
+            }
+            noteToEdit.setNodesSelected(originalNodes);
+        }
+
+        if (!transactionsSelected.isEmpty()) {
+            if (noteToEdit.isGraphAttribute()) {
+                noteToEdit.setGraphAttribute(false);
+            }
+            final List<Integer> originalTransactions = noteToEdit.getTransactionsSelected();
+            for (int transaction : transactionsSelected) {
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:794]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 8]
```diff
@@ -556,4 +730,107 @@ protected void closeEdit() {
...
+
+    /**
+     * Remove what is currently selected on the graph from the note's selected elements
+     */
+    public void removeFromSelectedElements(final NotesViewEntry noteToEdit) {
+        updateSelectedElements();
+
+        if (!nodesSelected.isEmpty()) {
+            final List<Integer> originalNodes = noteToEdit.getNodesSelected();
+            for (int node : nodesSelected) {
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:811]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java [hunk 9]
```diff
@@ -556,4 +730,107 @@ protected void closeEdit() {
...
+                    final int index = originalNodes.indexOf(node);
+                    originalNodes.remove(index);
+                }
+            }
+            noteToEdit.setNodesSelected(originalNodes);
+        }
+
+        if (!transactionsSelected.isEmpty()) {
+            final List<Integer> originalTransactions = noteToEdit.getTransactionsSelected();
+            for (int transaction : transactionsSelected) {
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/NotesViewPane.java:822]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewEntry.java [hunk 10]
```diff
@@ -29,12 +31,20 @@
...
-    public NotesViewEntry(final String dateTime, final String noteTitle, final String noteContent, final Boolean userCreated) {
+    public NotesViewEntry(final String dateTime, final String noteTitle, final String noteContent, final Boolean userCreated, final Boolean graphAttribute) {
         this.dateTime = dateTime;
         this.noteTitle = noteTitle;
         this.noteContent = noteContent;
         this.userCreated = userCreated;
+        this.graphAttribute = graphAttribute;
+        if (userCreated) {
+            this.nodesSelected = new ArrayList();
+            this.transactionsSelected = new ArrayList();
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewEntry.java:46]: diamond operators

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewEntry.java [hunk 11]
```diff
@@ -29,12 +31,20 @@
     private String noteTitle;
     private String noteContent;
     private final Boolean userCreated;
+    private Boolean graphAttribute;
+    private List<Integer> nodesSelected;
+    private List<Integer> transactionsSelected;
 
-    public NotesViewEntry(final String dateTime, final String noteTitle, final String noteContent, final Boolean userCreated) {
+    public NotesViewEntry(final String dateTime, final String noteTitle, final String noteContent, final Boolean userCreated, final Boolean graphAttribute) {
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewEntry.java:38]: use primitive `boolean`

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java [hunk 12]
```diff
@@ -60,12 +61,48 @@ public void readObject(final int attributeId, final int elementId, final JsonNod
...
+                                notesArray.get(i).get(0).asText(),
+                                notesArray.get(i).get(1).asText(),
+                                notesArray.get(i).get(2).asText(),
+                                notesArray.get(i).get(3).asBoolean(),
+                                notesArray.get(i).get(4).asBoolean()
+                        ));
+
+                        if (notesArray.get(i).get(3).asBoolean() == true && notesArray.get(i).get(4).asBoolean() == false) {
+                            // Add the selected nodes
+                            JsonNode nodesArrayNode = notesArray.get(i).get(5);
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java:75]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java [hunk 13]
```diff
@@ -60,12 +61,48 @@ public void readObject(final int attributeId, final int elementId, final JsonNod
...
+                            if (nodesArrayNode != null) {
+                                List<Integer> selectedNodes = new ArrayList();
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java:77]: final and diamond operator

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java [hunk 14]
```diff
@@ -60,12 +61,48 @@ public void readObject(final int attributeId, final int elementId, final JsonNod
...
+                                for (int j = 0; j < nodesArrayNode.size(); j++) {
+                                    selectedNodes.add(nodesArrayNode.get(j).asInt());
+                                }
+                                noteViewEntries.get(i).setNodesSelected(selectedNodes);
+                            }
+
+                            // Add the selected transactions
+                            JsonNode transactionsArrayNode = notesArray.get(i).get(6);
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java:85]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java [hunk 15]
```diff
@@ -60,12 +61,48 @@ public void readObject(final int attributeId, final int elementId, final JsonNod
...
+                            if (transactionsArrayNode != null) {
+                                List<Integer> selectedTransactions = new ArrayList();
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java:87]: final and diamond operator

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java [hunk 16]
```diff
@@ -109,12 +146,37 @@ public void writeObject(final Attribute attribute, final int elementId, final Js
                         jsonGenerator.writeString(note.getNoteTitle());
                         jsonGenerator.writeString(note.getNoteContent());
                         jsonGenerator.writeBoolean(note.isUserCreated());
+                        jsonGenerator.writeBoolean(note.isGraphAttribute());
+
+                        if (!note.isGraphAttribute()) {
+                            if (note.getNodesSelected() != null) {
+                                // Add nodes that are selected to the note
+                                final int nodesLength = note.getNodesSelected().size();
+                                int[] nodesArray = new int[nodesLength];
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java:155]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java [hunk 17]
```diff
@@ -109,12 +146,37 @@ public void writeObject(final Attribute attribute, final int elementId, final Js
...
+                                final List<Integer> nodesSelected = note.getNodesSelected();
+                                Iterator<Integer> nodesIterator = nodesSelected.iterator();
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java:157]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java [hunk 18]
```diff
@@ -109,12 +146,37 @@ public void writeObject(final Attribute attribute, final int elementId, final Js
...
+                                for (int i = 0; i < nodesLength; i++) {
+                                    nodesArray[i] = nodesIterator.next();
+                                }
+                                jsonGenerator.writeArray(nodesArray, 0, nodesLength);
+                            }
+
+                            if (note.getTransactionsSelected() != null) {
+                                // Add transactions that are selected to the note
+                                final int transactionsLength = note.getTransactionsSelected().size();
+                                int[] transactionsArray = new int[transactionsLength];
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java:167]: final

File: CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java [hunk 19]
```diff
@@ -109,12 +146,37 @@ public void writeObject(final Attribute attribute, final int elementId, final Js
...
+                                final List<Integer> transactionsSelected = note.getTransactionsSelected();
+                                Iterator<Integer> transactionsIterator = transactionsSelected.iterator();
```
Comment (antares1470) [inline CoreNotesView/src/au/gov/asd/tac/constellation/views/notes/state/NotesViewStateIoProvider.java:169]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]
If you're able to include a whatsnew entry as well as some changelog entries for the edited function definitions, that would be great

**Review by antares1470 — CHANGES_REQUESTED** [review]
On top of the above, I tested the branch and encountered the following things:
- If I untick 'Apply note to graph selection', the note appears to be lost (can't see the note even with User Notes filter on)
- With 'Apply note to graph selection' ticked, it appears the note doesn't always apply to the selection. I created a note with just one node selected and sometimes the note was applied to the node but other times it applied to the whole graph and other times still applied to none of the graph
- I observed that if you choose the Selected filter that it deselects the other two. I'm ok with this but I noted that you can't then select any of the other filters until you untick the Selected filter which I don't think is intuitive. Maybe if you select one of the other filters it deselects the Selected filter?

**Comment by Delphinus8821** [comment]
Example of the added description for user notes about what graph elements are linked to the note (Line at the bottom of each note). 

AddedDescription

**Review by arcturus2 — APPROVED** [review]
Tested and works well. The issues @antares1470 mentioned seem to be fixed.

**Review by antares1470 — CHANGES_REQUESTED** [review]
Tested again and have a couple of things:
- If I have nothing selected on my graph and have the Apply to graph selection checkbox ticked, the note I create is applied to 0 nodes and 0 transactions and so can only be seen through the User Notes filter. While this behaviour is technically working, I think it would be cool to treat no selection the same as applying to the whole graph (similar to arrangements)

- I created a graph with 2 nodes and a transaction and then created 2 notes: one applied to the whole graph and one applied to just a single node. When I used the right-click context menu on the whole graph note, it selected the whole graph. When I then did the same for the other note, it deselected the transaction but not the node (that the note wasn't applied to). I think it should deselect everything on the graph that the note was applied to (similar to Select on graph in QCV)

**Review by antares1470 — APPROVED** [review]

**Pull request merged**